### PR TITLE
feat: Add tax for tickets

### LINKED ISSUE #2120: Fetch tex price details
**Describe the feature you'd like**
Display tex price with the ticket price

**Would you like to work on the issue?**

yes

### PR DESCRIPTION
**Pull request by liveHarshit**
Fixes #2120 

<p align="center">
 
 
</p>

### CODE REVIEW

**Comment by mariobehling** [comment]
Thanks. Please also provide an example for countries where tax is included in the price. Check how this works in Europe on Eventbrite for example and also check the frontend, please.

**Comment by liveHarshit** [comment]
> Thanks. Please also provide an example for countries where tax is included in the price. Check how this works in Europe on Eventbrite for example and also check the frontend, please.

Yes, I have checked on the frontend. It is according to organizer preference and all information are fetching from the server. 
image
image

File: app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt [hunk 1]
```diff
@@ -92,4 +94,6 @@ abstract class OpenEventDatabase : RoomDatabase() {
     abstract fun notificationDao(): NotificationDao
 
     abstract fun settingsDao(): SettingsDao
+
+    abstract fun texDao(): TaxDao
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt]: texDao ?

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt]: It should be taxDao*, will update it :+1:

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt]: updated

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt [hunk 2]
```diff
@@ -37,7 +41,11 @@ class TicketsViewModel(
     val event: LiveData<Event> = mutableEvent
     private val mutableDiscountCodes = MutableLiveData<DiscountCode>()
     val discountCode: LiveData<DiscountCode> = mutableDiscountCodes
+    private val mutableTaxInfo = MutableLiveData<Tax>()
+    val taxInfo: LiveData<Tax> = mutableTaxInfo
     var appliedDiscountCode: DiscountCode? = null
+    var tax: Tax? = null
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Remove this variable. Totally unnecessary

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: removed

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt [hunk 3]
```diff
@@ -122,23 +130,44 @@ class TicketsViewModel(
                 val code = appliedDiscountCode
                 tickets.forEach { ticket ->
                     var price = ticket.price
+                    totalTaxAmount += (ticket.price * taxRate / 100) * qty[index]
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Why are we not using it.second instead of maintaining an index?

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketViewHolder.kt [hunk 4]
```diff
@@ -28,6 +29,7 @@ import kotlinx.android.synthetic.main.item_ticket.view.description
 import org.fossasia.openevent.general.discount.DiscountCode
 import org.fossasia.openevent.general.event.EventUtils
 import org.fossasia.openevent.general.event.EventUtils.getFormattedDate
+import org.fossasia.openevent.general.event.tex.Tax
```
Comment (anhanh11001) [inline app/src/main/java/org/fossasia/openevent/general/ticket/TicketViewHolder.kt]: folder name should be tax, not tex

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt [hunk 5]
```diff
@@ -122,23 +130,44 @@ class TicketsViewModel(
...
                     if (code?.value != null) {
                         appliedDiscountCode?.tickets?.forEach { ticketId ->
                             if (ticket.id == ticketId.id.toInt()) {
                                 price -= if (code.type == AMOUNT) code.value else price*(code.value / 100)
                             }
                         }
                     }
-                    price.let { prices += price * qty[index++] }
+                    price.let { prices += price * qty[index] }
```
Comment (anhanh11001) [inline app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Unnecessary change?

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Not unnecessary, the index is updated at the end.

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt (hunk 3 above)
Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: it is nothing here. We are fetching the list of tickets and qty stored in an array.

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketViewHolder.kt (hunk 4 above)
Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketViewHolder.kt]: changed

File: app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt (hunk 3 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Why not use ticketAndQty?

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: I think so. It's quite complex. Will try :+1:

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: It's not. it was created exactly so that index can be avoided

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: ticketAndQty is a list, so the index is required for this also. `ticketAndQty[index].second`

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: No, it is a list of pair. So you iterate over it without the need of index

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Iterating is already used on a list of tickets which is getting from the database for the ticket price. So for this, two parallel iterators are required.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Don't use tickets for iteration, use ticketAndQty, that's why it was created by @anhanh11001

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Well then ticketAndQty doesn't have the ticket price
```
E/TicketsViewModel: (47, 1, 0.0)
E/TicketsViewModel: (48, 2, 0.0)
```
So here we need to fetch selected tickets to get price. Only donation tickets have price at third place in ticketAndQty.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/ticket/TicketsViewModel.kt]: Merging, but open an issue to either use or remove ticketAndQty
